src/native/libs/System.Security.Cryptography.Native/pal_evp_kem.c

```diff
@@ -0,0 +1,301 @@
+
+    return NULL;
+}
+
+int32_t CryptoNative_EvpKemEncapsulate(EVP_PKEY* pKey,
+                                       uint8_t* ciphertext,
+                                       int32_t ciphertextLength,
+                                       uint8_t* sharedSecret,
+                                       int32_t sharedSecretLength)
+{
+    assert(pKey);
+    assert(ciphertext);
+    assert(sharedSecret);
+
+#ifdef NEED_OPENSSL_3_0
+    if (API_EXISTS(EVP_PKEY_encapsulate_init))
+    {
+        assert(API_EXISTS(EVP_PKEY_CTX_new_from_pkey) && API_EXISTS(EVP_PKEY_encapsulate));
+        ERR_clear_error();
+
+        EVP_PKEY_CTX* ctx = NULL;
+        ctx = EVP_PKEY_CTX_new_from_pkey(NULL, pKey, NULL);
+        int32_t ret = 0;
+
+        if (ctx == NULL)
+        {
+            goto done;
+        }
+
+        if (EVP_PKEY_encapsulate_init(ctx, NULL) != 1)
+        {
+            goto done;
+        }
+
+        size_t ciphertextLengthT = Int32ToSizeT(ciphertextLength);
+        size_t sharedSecretLengthT = Int32ToSizeT(sharedSecretLength);
+
+        if (EVP_PKEY_encapsulate(ctx, ciphertext, &ciphertextLengthT, sharedSecret, &sharedSecretLengthT) != 1)
+        {
+            goto done;
+        }
+
+        if (ciphertextLengthT != Int32ToSizeT(ciphertextLength) || sharedSecretLengthT != Int32ToSizeT(sharedSecretLength))
+        {
+            ret = -1;
+        }
+
+        ret = 1;
+
+done:
+        if (ctx != NULL)
+        {
+            EVP_PKEY_CTX_free(ctx);
+        }
+
+        return ret;
+    }
+#endif
+
+    (void)pKey;
+    (void)ciphertext;
+    (void)ciphertextLength;
+    (void)sharedSecret;
+    (void)sharedSecretLength;
+    return 0;
+}
+
+int32_t CryptoNative_EvpKemDecapsulate(EVP_PKEY* pKey,
+                                       const uint8_t* ciphertext,
+                                       int32_t ciphertextLength,
+                                       uint8_t* sharedSecret,
+                                       int32_t sharedSecretLength)
+{
+    assert(pKey);
+    assert(ciphertext);
+    assert(sharedSecret);
+
+#ifdef NEED_OPENSSL_3_0
+    if (API_EXISTS(EVP_PKEY_decapsulate_init))
+    {
+        assert(API_EXISTS(EVP_PKEY_CTX_new_from_pkey) && API_EXISTS(EVP_PKEY_decapsulate));
+        ERR_clear_error();
+
+        EVP_PKEY_CTX* ctx = NULL;
+        ctx = EVP_PKEY_CTX_new_from_pkey(NULL, pKey, NULL);
+        int32_t ret = 0;
+
+        if (ctx == NULL)
+        {
+            goto done;
+        }
+
+        if (EVP_PKEY_decapsulate_init(ctx, NULL) != 1)
+        {
+            goto done;
+        }
+
+        size_t ciphertextLengthT = Int32ToSizeT(ciphertextLength);
+        size_t sharedSecretLengthT = Int32ToSizeT(sharedSecretLength);
+
+        if (EVP_PKEY_decapsulate(ctx, sharedSecret, &sharedSecretLengthT, ciphertext, ciphertextLengthT) != 1)
+        {
+            goto done;
+        }
+
+        if (sharedSecretLengthT != Int32ToSizeT(sharedSecretLength))
+        {
+            ret = -1;
```

Comment:
isn't this no-op given line below you set it back to 1

Reply:
That was missing a `goto done`. Thanks for the catch!